### ISSUE TITLE
OpenGL ES 2.0 + EGL port

### ISSUE THREAD

**Issue by tapio** [performous/performous#17]
The new rendering engine in 0.7.0 should make this rather straight-forward. Implementing this would allow us to try porting the game to new embedded and mobile platforms such as the Raspberry Pi (which has been requested a few times; performance might become an obstacle but it's worth a try).

**Comment by fpoirotte** [comment]
So... the new raspberry 2 type B is now out and comes with more RAM, a faster CPU and ARM7 support.
Any chance to see another go at this issue?
Like others here, I'd love to be able to play around with performous on a rpi.

**Comment by TobiX** [comment]
As I already own a Raspberry Pi 2, I will play around with this. The switch from GLEW to libepoxy was the first step to get this working without too much "extra" code.

**Comment by wellumies** [comment]
Could this port work on http://www.banana-pi.org/m3.html ?

**Comment by TobiX** [comment]
If the GL stack isn't broken like on most distributions I tried last year on the Raspberry Pi 2, you might want to give it a try. There are  probably still a lot of places in the code to fix and the shaders might need some tuning...

**Comment by nieknooijens** [comment]
I tested it a month ago on the last raspbian but the RPi2 openGL driver is very, very unstable, unfortunately we need it because performous depends on GLX.

**Comment by TobiX** [comment]
Yes, but the GLX dependency is very light nowadays (I think). We don't link against GLX directly, but to even start squashing  the bugs in our GLES suport, you need to disable webcam support, since OpenCV links to libGL directly...

**Comment by nieknooijens** [comment]
obsolete since we use libepoxy and raspbian offers experimental "fat" openGL support.

**Comment by TobiX** [comment]
We still have no proper EGL support...

**Comment by nieknooijens** [comment]
@TobiX it runs on a Pi2 using libepoxy, see: https://www.dropbox.com/s/7o8ar3lumqzlsb4/VID_20170124_100548.mp4?dl=0

**Comment by tatankat** [comment]
Tried running on a Pi3:
- I have (latest master without webcam) running under X, using OpenGL, but I had to compile libSDL2 (2.0.5) without opengles. Another bug is fixed using master checkout of libepoxy. There are still some minor bugs present, but it works, although it constantly fully uses all cores.
- running with libSDL2 from repository (using performous from repo and also latest from git) under X gives me: "SDL_SetVideoMode failed: Could not get EGL display"
- running from tty, using libSDL2 from repository (with EGL support I assume), gives me "OpenGL 2.1 is required but not available".
   - After disabling this check: "performous: Couldn't find current GLX or EGL context."
   - Using libepoxy from master it gives me: "FATAL ERROR: core.vert: Couldn't create shader."

**Comment by TobiX** [comment]
> Using libepoxy from master it gives me: "FATAL ERROR: core.vert: Couldn't create shader."

That is the missing GLES support in Performous' code (which would also be nice for Android support...) - Maybe someone with more GL experience then me could take a look at the shaders and make them compatible with GLES... The egl branch might have some hints.

(Be aware that "fat" OpenGL support on RPi is not much better then a software renderer, since it falls back to software rendering for many operations - alas, it might be enough for Performous...)

**Comment by nieknooijens** [comment]
to be honest. is EGL still a thing? the nvidia shield and some other android devices can already run "fat openGL" and the Raspberry pi 2 can run it as well. isn't it a better idea to port to vulkan instead?

**Comment by mirabilos** [comment]
I just found https://github.com/p3/regal which might help with this and perhaps https://github.com/performous/performous/issues/482 as it apparently can emulate OpenGL on OpenGL/GL ES, without the full performance drag from emulation (6–8 fps on Thinkpad X61, Core2duo 2 GHz, with llvmpipe software rendering).

Might be useful to try whether it works, perhaps even without code changes (`LD_PRELOAD` is mentioned).

**Comment by Baklap4** [comment]
Reviving the old thread
I don't own a Raspberry so can't quite test it.
Current status is that it still doesn't work as mentioned in discord

Build + install went fine, but then a runtime failure appeared:
```
video/warning: Error setting GLattr m_attr: requested 1, got 0
video/warning: Error setting GLattr m_attr: requested 24, got 0
video/warning: Error setting GLattr m_attr: requested 32, got 0
video/warning: Error setting GLattr m_attr: requested 8, got 0
video/warning: Error setting GLattr m_attr: requested 8, got 0
video/warning: Error setting GLattr m_attr: requested 8, got 0
video/warning: Error setting GLattr m_attr: requested 8, got 0
video/warning: Error setting GLattr m_attr: requested 1, got 0
core/notice: More details might be available in "/home/jl/.cache/performous/infolog.txt".
logger/info: Exiting normally.
FATAL ERROR: SDL_GL_CreateContext failed with error: Could not create GL context: GLXBadFBConfig
core/error: SDL_GL_CreateContext failed with error: Could not create GL context: GLXBadFBConfig
```

**Comment by izar** [comment]
@Baklap4 try running with MESA_GL_VERSION_OVERRIDE=4.3 - works for me on a RP4 running Raspbian. 

Sound is still borked, though.

**Comment by ooshlablu** [comment]
@izar That's great to know!
What do you mean that the sound is borked? Are there sound devices messed up, or no output over HDMI, or something similar? Do you happen to have like a Behringer or MobilePRE USB interface you could try it with?
I won't be able to get to it for a few days, but I will give this a try also. I end up using a Behringer Uphoria most of the time anyway on my x86 hardware. Will keep you posted

**Comment by izar** [comment]
hi @ooshlablu - sorry for the brevity of the description. I was going between ultradx, pykaraoke and performous. Got to a point where ultradx works, after some juggling of alsa and pulseaudio configs. Performous, unfortunately, exits with alsa error messages that I cannot make go away.
I am not close to my RP4 right now, but as soon as I am I'll try to collect the exact information and open a different ticket just for it.
